public/js/particle_manager.js

```diff
@@ -0,0 +1,39 @@
+import {Particle} from "./particle.js";
+import {random, randomChoice} from "./util.js";
+
+export class ParticleManager {
+
+    constructor(options) {
+        this.particles = [];
+        this.particlesIndex = 0;
+        Object.assign(this, options);
+        //preload all the particle objects, so they can be reused.
+        this.increaseParticleCountTo(this.particlesMax);
+    }
+
+    increaseParticleCountTo(number = 0) {
+        this.particlesMax = Math.max(number, this.particlesMax);
+        // only create new objects if absolutely necessary - reuse otherwise.
+        for (let i = this.particles.length; i < number; i++) 
```

Comment:
Fulfils requirement 8 by reusing objects for particles (up to 100 objects per second were being created).